### PR TITLE
🔧 Dynamic Consumer Groups

### PR DESCRIPTION
**Pull request by awtkns**
- Allows devs to share a single kafka cluster
- In dev mode, you will only consume message you produce
- Each person on the cluster will be assinged thier own consumer group
- Messages will be skipped if it is not produced by your maching
- Each consumer group keeps track of its own oftset

### CODE REVIEW

File: platform/reworkd_platform/services/kafka/consumers/base.py [hunk 1]
```diff
@@ -50,11 +53,15 @@ async def start(self) -> "AsyncConsumer":
...
-            try:
-                async for msg in consumer:
+            async for msg in consumer:
+                if self.should_skip(msg):
+                    logger.info(f"Skipping message: {msg.headers}")
+                    continue
+
+                try:
                     await self.on_message(msg)
-            finally:
```
Comment (awtkns) [inline platform/reworkd_platform/services/kafka/consumers/base.py:56]: this should remain

File: platform/reworkd_platform/services/kafka/consumers/base.py [hunk 2]
```diff
@@ -25,22 +25,25 @@ def __init__(
...
         settings: Settings,
-        deserializer: Deserializer = JSONDeserializer()
+        deserializer: Deserializer = JSONDeserializer(),
     ):
+        self._group = settings.kafka_consumer_group
         self._consumer = settings.kafka_enabled and AIOKafkaConsumer(
             *topics,
             bootstrap_servers=settings.kafka_bootstrap_servers,
-            group_id="platform",
+            group_id=self._group,
```
Comment (asim-shrestha) [inline platform/reworkd_platform/services/kafka/consumers/base.py:34]: Suggestion:
```
        self._consumer = settings.kafka_enabled and AIOKafkaConsumer(
            *topics,
            bootstrap_servers=settings.kafka_bootstrap_servers,
            group_id=settings.kafka_consumer_group,
```

File: platform/reworkd_platform/services/kafka/consumers/base.py [hunk 3]
```diff
@@ -50,11 +53,15 @@ async def start(self) -> "AsyncConsumer":
...
+                if self.should_skip(msg):
+                    logger.info(f"Skipping message: {msg.headers}")
+                    continue
+
+                try:
                     await self.on_message(msg)
-            finally:
-                await self.stop()
+                except Exception as e:
+                    logger.exception(e)
```
Comment (asim-shrestha) [inline platform/reworkd_platform/services/kafka/consumers/base.py:64]: So will this stop this thing from crashing on every exception?

File: platform/reworkd_platform/services/kafka/consumers/base.py [hunk 4]
```diff
@@ -65,6 +72,12 @@ async def stop(self) -> None:
 
         await self._consumer.stop()
 
+    def should_skip(self, record: ConsumerRecord) -> bool:
+        return (
+            self._dev_mode
+            and ("host", self._group.encode("utf-8")) not in record.headers
+        )
```
Comment (asim-shrestha) [inline platform/reworkd_platform/services/kafka/consumers/base.py:79]: Can you explain this a bit?

File: platform/reworkd_platform/settings.py [hunk 5]
```diff
@@ -102,6 +108,13 @@ class Settings(BaseSettings):
     ff_mock_mode_enabled: bool = False  # Controls whether calls are mocked
     max_loops: int = 25  # Maximum number of loops to run
 
+    @property
+    def kafka_consumer_group(self) -> str:
+        if self.environment == "development":
+            return platform.node()
+
+        return "platform"
```
Comment (asim-shrestha) [inline platform/reworkd_platform/settings.py:116]: what is platform.node()?

File: platform/reworkd_platform/services/kafka/consumers/base.py (hunk 3 above)
Comment (awtkns) [reply platform/reworkd_platform/services/kafka/consumers/base.py:64]: yes

**Pull request merged**